pkg/ruler/ruler_test.go

```diff
@@ -216,3 +228,373 @@ func compareRuleGroupDescToStateDesc(t *testing.T, expected *rules.RuleGroupDesc
+				}
+			},
+
+			expectedRules: expectedRulesMap{
+				// This ruler doesn't get rules from unhealthy ruler (RF=1).
+				ruler1: map[string]rules.RuleGroupList{
+					user1: {user1Group1},
+					user2: {user2Group1},
+				},
+				ruler2: noRules,
+			},
+		},
+
+		"default sharding, LEAVING ruler": {
+			sharding:         true,
+			shardingStrategy: ShardingStrategyDefault,
+
+			setupRing: func(desc *ring.Desc) {
+				desc.AddIngester(ruler1, ruler1Addr, "", sortTokens([]uint32{user1Group1Token + 1, user2Group1Token + 1}), ring.LEAVING)
+				desc.AddIngester(ruler2, ruler2Addr, "", sortTokens([]uint32{user1Group2Token + 1, user3Group1Token + 1}), ring.ACTIVE)
+			},
+
+			expectedRules: expectedRulesMap{
+				// LEAVING ruler doesn't get any rules.
+				ruler1: noRules,
+				ruler2: allRules,
+			},
+		},
+
+		"default sharding, JOINING ruler": {
+			sharding:         true,
+			shardingStrategy: ShardingStrategyDefault,
+
+			setupRing: func(desc *ring.Desc) {
+				desc.AddIngester(ruler1, ruler1Addr, "", sortTokens([]uint32{user1Group1Token + 1, user2Group1Token + 1}), ring.JOINING)
+				desc.AddIngester(ruler2, ruler2Addr, "", sortTokens([]uint32{user1Group2Token + 1, user3Group1Token + 1}), ring.ACTIVE)
+			},
+
+			expectedRules: expectedRulesMap{
+				// JOINING ruler has no rules yet.
+				ruler1: noRules,
+				ruler2: allRules,
+			},
+		},
+
+		"shuffle sharding, single ruler": {
+			sharding:         true,
+			shardingStrategy: ShardingStrategyShuffle,
+
+			setupRing: func(desc *ring.Desc) {
+				desc.AddIngester(ruler1, ruler1Addr, "", sortTokens([]uint32{0}), ring.ACTIVE)
+			},
+
+			expectedRules: expectedRulesMap{
+				ruler1: allRules,
+			},
+		},
+
+		"shuffle sharding, multiple rulers, shard size 1": {
+			sharding:         true,
+			shardingStrategy: ShardingStrategyShuffle,
+			shuffleShardSize: 1,
+
+			setupRing: func(desc *ring.Desc) {
+				desc.AddIngester(ruler1, ruler1Addr, "", sortTokens([]uint32{userToken(user1, 0) + 1, userToken(user2, 0) + 1, userToken(user3, 0) + 1}), ring.ACTIVE)
+				desc.AddIngester(ruler2, ruler2Addr, "", sortTokens([]uint32{user1Group1Token + 1, user1Group2Token + 1, user2Group1Token + 1, user3Group1Token + 1}), ring.ACTIVE)
+			},
+
+			expectedRules: expectedRulesMap{
+				ruler1: allRules,
+				ruler2: noRules,
+			},
+		},
+
+		// Same test as previous one, but with shard size=2. Second ruler gets all the rules.
+		"shuffle sharding, two rulers, shard size 2": {
+			sharding:         true,
+			shardingStrategy: ShardingStrategyShuffle,
+			shuffleShardSize: 2,
+
+			setupRing: func(desc *ring.Desc) {
+				desc.AddIngester(ruler1, ruler1Addr, "", sortTokens([]uint32{userToken(user1, 0) + 1, userToken(user2, 0) + 1, userToken(user3, 0) + 1}), ring.ACTIVE)
+				desc.AddIngester(ruler2, ruler2Addr, "", sortTokens([]uint32{user1Group1Token + 1, user1Group2Token + 1, user2Group1Token + 1, user3Group1Token + 1}), ring.ACTIVE)
```

Comment:
I'm confused why there are specific rule group tokens in the ring setup for this test.

Reply:
This test uses the exact same tokens setup as previous test `"shuffle sharding, multiple rulers, shard size 1"`, but with shard size=2. I've added comment about it.

Reply:
👍